Dockerfile

```diff
@@ -0,0 +1,27 @@
+# Build the manager binary
+FROM golang:1.12.5 as builder
+
+WORKDIR /workspace
+# Copy the Go Modules manifests
+COPY go.mod go.mod
+COPY go.sum go.sum
+# cache deps before building and copying source so that we don't need to re-download as much
+# and so that source changes don't invalidate our downloaded layer
+RUN go mod download
+
+# Copy the go source
+COPY main.go main.go
+COPY api/ api/
+COPY controllers/ controllers/
+
+# Build
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
+
+# Use distroless as minimal base image to package the manager binary
+# Refer to https://github.com/GoogleContainerTools/distroless for more details
+FROM gcr.io/distroless/static:nonroot 
```

Comment:
personally not keen on `distroless` as it can make debugging quite challenging if you lack observability. perhaps the solution is to have better observability but that's not a satisfying response in the middle of a 4am incident. could we use something like alpine which at least contains a shell + package manager? anyone else have preferences here?

Reply:
> …make debugging quite challenging if you lack observability…
> …contains a shell + package manager…

What do you find a shell in the container useful for during live debugging?

I'd generally go for as light as possible, both to reduce download size and minimise attack surface. But I'm not opposed to Alpine configured to run as non-root if we feel that directly shelling onto the container is going to be useful.

Reply:
> What do you find a shell in the container useful for during live debugging?

mostly debugging network issues. anything from "is this hostport accessible from inside the container", checking that dns resolves correctly in the container, or even grabbing full tcpdumps for some nuance issues.  this is very difficult if your container doesn't have these tools preinstalled, or access to a busybox

> as light as possible

distroless tends to be larger than alpine

Reply:
Personally I'd definitely lean towards distroless, but can appreciate concerns around debuggability. On the other hand..

1) I'd imagine there'll not really be anything to debug within the operators controller - it exclusively talks to external APIs, so I think the only really debug surface will be log messages.

2) 'debug containers' are coming in Kubernetes, and have been released in alpha in 1.16

Distroless is pretty much scratch + ca-certificates, so in terms of 'surface' I think it's a lot smaller.

Given how easy it'll be to change this in future though, I don't think it's too much of an issue to use alpine. But I'd imagine we'll want to switch to distroless ahead of release 😄

Reply:
> 'debug containers' are coming in Kubernetes

awesome! can't wait

> But I'd imagine we'll want to switch to distroless ahead of release

agreed